---
title: "Anopheles gambiae spatial genome analysis"
author: "Clara Rehmann"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE, results="hide"}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/phd/research/spatial_sweeps/anopheles")
library(tidyverse)
library(data.table)
library(MASS)
library(patchwork)
library(ggiraph)
library(scales)
library(quantreg)
library(viridis)
library(splines)
library(stats)
library(npreg)
library(ggnewscale)
load("../../../locator/data/cntrymap.Rdata")
```

# Anopheles dataset

```{r metadata, echo=FALSE}
metadata <- fread('data/admixture_k1_metadata.txt')
metadata_count <- metadata %>% group_by(x, y) %>% count()
```

After running Admixture, we identified three distinct populations of *Anopheles* in the Ag1000g dataset. For our genome scan analysis, we'll focus on the `nrow(metadata)` individuals that have at least 90% ancestry assigned to population **K1**, which 
contains the most unique sampling locations and covers the largest area of the landscape.

```{r sampling-locs, echo=FALSE}
ggplot()+coord_map(projection = "mollweide",
                         xlim=c(min(na.omit(c(metadata$x)))-1,
                                max(na.omit(c(metadata$x)))+1),
                         ylim=c(min(na.omit(c(metadata$y)))-1,
                                max(na.omit(c(metadata$y)))+1))+
        theme_classic()+
        geom_polygon(data=fortify(map),aes(x=long,y=lat,group=group),
                     fill="snow2",color="white",lwd=0.2)+
        geom_count(data=metadata_count,aes(x=x,y=y,fill=n,size=n),color='black',pch=21) +
        scale_fill_viridis(option='mako', trans='log10', end=0.9)
```

## Spatial sampling considerations

The studies that contributed these samples don't all use the exact same sampling approaches, so I looked into 
how sampling areas were identified for each study to inform how I'd calculate the spatial spread of a given allele. 
The spatially-proximate samples in Mali and Cameroon are from studies that sample transects, and take care in the design to 
not have spatial overlap with their sampling points (ie, the areas being sampled should have distinct populations). 
Beyond that, the approximate area sampled in a given study seems to be around 1-10 km2 (probably closer to 1 km2).
To be safe, I'll only include alleles that are observed at at least three locations.
To approximate the area this allele occupies, 
I'll fit a convex hull over the space and report the area of that shape.

# Spatial genome scan

```{r process-scan, eval=FALSE}
# code block for reading in and compiling all windowed spatial genome scan files
rf <- function(filepath){
  df <- fread(filepath)
  df$chromosome <- strsplit(filepath, '[.]')[[1]][1] %>% str_remove('out/windowed-genome-scan//')
  return(df)
}

allsnps <- rbindlist(lapply(list.files('out/windowed-genome-scan/', pattern='_spatial_genome_scan.txt', full.names=TRUE), rf), fill=TRUE)
allsnps <- allsnps[allsnps$locs > 2,]

# annotate SNPs in inversions
allsnps$INV <- 'none'

# start and end positions for each inversion feature
Rbstart <- 18575300
Rbstop <- 26767588
allsnps[(allsnps$chromosome=='2R' &
    allsnps$position >= Rbstart &
    allsnps$position <= Rbstop),]$INV <- '2Rb'
Rcstart <- 26750000
Rcstop <- 31473000
allsnps[(allsnps$chromosome=='2R' &
    allsnps$position >= Rcstart &
    allsnps$position <= Rcstop),]$INV <- '2Rc'
Lastart <- 20524058
Lastop <- 42165532
allsnps[(allsnps$chromosome=='2L' &
    allsnps$position >= Lastart &
    allsnps$position <= Lastop),]$INV <- '2La'


# add associated gene to each SNP...

aggenes <- fread('data/agPEST_genelist.txt')
aggenes <- aggenes %>% tidyr::separate(col=`Genomic Location (Transcript)`,
                     into=c(NA,'start',NA,'stop',NA),
                     sep='[:|..|(]',
                     convert=TRUE)

allsnps$`Gene ID` <- 'none'
allsnps$`Product Description` <- 'none'

addgene <- function(allsnps, chromosome_name, start_position, stop_position, geneID, productDescription, sourceID){
                    allsnps[allsnps$chromosome == chromosome_name & allsnps$position >= start_position & allsnps$position <= stop_position,'Gene ID'] <- geneID
                    allsnps[allsnps$chromosome == chromosome_name & allsnps$position >= start_position & allsnps$position <= stop_position,'Product Description'] <- productDescription
                    return(allsnps)
}

for (i in seq(nrow(aggenes))){
  chromosome_name=aggenes[i,'Chromosome']
  start_position=aggenes[i,'start']
  stop_position=aggenes[i,'stop']
  geneID=aggenes[i,'Gene ID']
  productDescription=aggenes[i,'Product Description']
  sourceID=aggenes[i,'source_id']
  
  allsnps <- addgene(allsnps, chromosome_name, start_position, stop_position, geneID, productDescription, sourceID)
  print(i)
}

allsnps <- allsnps[allsnps$area > 0,]

fwrite(allsnps, 'out/all_snps_annotated.txt', sep='\t')
```

# Joint frequency-area distribution

xyz about frequency vs area

```{r read-data, echo=FALSE}
allsnps <- fread('out/all_snps_annotated.txt')
ggplot(allsnps) + theme_bw() + 
  geom_point(aes(x=frequency, y=area), alpha=0.005)
```

To identify outliers, we should apply our method separately to
each inversion and then the rest of the genome, since the inversions
move weirdly with regards to space and frequency and thus could mess with things.

```{r separate-inv}
invsnps <- allsnps[allsnps$INV != 'none',]
allsnps <- allsnps[allsnps$INV == 'none',]
```

## Outlier SNPs

### Genomic

Now, let's find outliers in the joint distribution using the 
lower tenth percentile of the distribution as the outlier cutoff:

```{r find-outliers}
# bin frequencies by every N alleles
N = 1000
allsnps <- allsnps %>% arrange(by = frequency)
bins <- rep(seq(from = 0, to=as.integer(nrow(allsnps)/ N)), each=N)[1:nrow(allsnps)]


allsnps$bin <- bins
allsnps <- allsnps %>% group_by(bin) %>% mutate(bin_frequency = mean(frequency))

# lower Q quantile for each frequency bin
Q = 0.05
allsnps <- allsnps %>% dplyr::group_by(bin) %>% dplyr::mutate(Q = quantile(area, probs=0.1))

# try fitting a spline?
m_reml <- ss(allsnps$frequency, allsnps$Q, method='REML', spar=0.4)
p_reml <- predict(m_reml, allsnps$frequency)
allsnps$cutoff <- p_reml$y
allsnps$outlier <- FALSE
allsnps[allsnps$area <= allsnps$cutoff,]$outlier <- TRUE

# grab outliers to a dataframe
outliersnps <- allsnps[allsnps$outlier == TRUE,]
fwrite(outliersnps, 'out/outlier_snps_annotated.txt', sep='\t')

ggplot(allsnps) + geom_point(aes(x=frequency, y=area, color=outlier), alpha=0.1) + geom_line(aes(x=frequency, y=cutoff))
```

Let's also give each SNP a bin-based Z score based on where it is w/in the bulk of the distribution:

```{r snprob}
robustscale <- function(data, center=TRUE, scale=TRUE,
                        preserveScale = FALSE){
  medians = NULL
  if(center){
    medians <- median(data,na.rm=TRUE)
    data = data-medians
  }
  mads=NULL
  if(scale){
    mads <- mad(data, na.rm=TRUE)
    if (mads==0){mads <- 1e-6}
    if(preserveScale){
      mads <- mads/mean(mads)
    }
    data = data/mads
  }
  return(data)
}

allsnps <- allsnps %>% dplyr::group_by(bin) %>% dplyr::mutate(zrob = robustscale(area))
```


```{r zcSNPwise}
allsnps <- allsnps %>% dplyr::group_by(bin) %>% dplyr::mutate(binmean = mean(area), binsd = sd(area))
allsnps <- allsnps %>% dplyr::mutate(snpZscore = (area - binmean)/binsd)
```

Are certain types of SNPs more likely to be found amongst outliers?
Let's check using a hypergeometric test:

** I'm not sure how to get enrichment from this or if I'm calculating it right! check with peter :) **

```{r hypergeometric, echo=FALSE}
for (ann in unique(allsnps$annotation)){
  q <- nrow(allsnps[allsnps$outlier == TRUE & allsnps$annotation == ann,])
  m <- nrow(allsnps[allsnps$annotation == ann,])
  n <- nrow(allsnps[! allsnps$annotation == ann,])
  k <- nrow(allsnps[allsnps$outlier == TRUE,])
  
  print(paste(ann, ': p = ', phyper(q, m, n, k, lower.tail=FALSE)))
}
```

Are the genes that outlier SNPs are in enriched for any sorts of functions? I'll output Ensembl gene IDs so I can run DAVID for enrichment tests:

```{r outlier-snp-genes-DAVID}
NSoutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$annotation == 'missense_variant',]$`Gene ID`
NSoutliergenes <- as.list(unique(NSoutliergenes[NSoutliergenes != 'none']))
fwrite(NSoutliergenes, 'out/NS_outlier_genes.txt', sep='\n')

SYoutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$annotation == 'synonymous_variant',]$`Gene ID`
SYoutliergenes <- as.list(unique(SYoutliergenes[SYoutliergenes != 'none']))
fwrite(SYoutliergenes, 'out/SY_outlier_genes.txt', sep='\n')

outliergenes <- allsnps[allsnps$outlier == TRUE,]$`Gene ID`
outliergenes <- as.list(unique(outliergenes[outliergenes != 'none']))
fwrite(outliergenes, 'out/outlierSNP_genes.txt', sep='\n')

hifioutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$frequency >= 0.75,]$`Gene ID`
hifioutliergenes <- as.list(unique(hifioutliergenes[hifioutliergenes != 'none']))
fwrite(hifioutliergenes, 'out/hifioutlierSNP_genes.txt', sep='\n')
```

### Inversions

Now, I'll do the same for each of the inversions:

```{r inversion-outliers, echo=FALSE, results='asis'}
dir.create('out/inversion', showWarnings = FALSE)
Rbstart <- 18575300
Rbstop <- 26767588
Rcstart <- 26750000
Rcstop <- 31473000
Lastart <- 20524058
Lastop <- 42165532
for (inv in c('2Rb','2Rc','2La')){
  cat(paste0("#### ",inv),'\n\n')
  invdf <- invsnps[invsnps$INV == inv,]
  
  #### find outliers
  
  # bin frequencies by every N alleles
  N = 50
  invdf <- invdf %>% arrange(by = frequency)
  bins <- rep(seq(from = 0, to=as.integer(nrow(invdf)/ N)), each=N)[1:nrow(invdf)]
  invdf$bin <- bins
  # 10th percentile cutoff
  invdf <- invdf %>% dplyr::group_by(bin) %>%
                     dplyr::mutate(Q = quantile(area, probs=0.1))
  # fit spline to ID outlier cutoff
  m_reml <- ss(invdf$frequency, invdf$Q, method='REML', spar=0.4)
  p_reml <- predict(m_reml, invdf$frequency)
  invdf$cutoff <- p_reml$y
  # ID outliers
  invdf$outlier <- FALSE
  invdf[invdf$area <= invdf$cutoff,]$outlier <- TRUE
  # plot it
  print(ggplot(invdf) + geom_point(aes(x=frequency, y=area, color=outlier), alpha=0.1) +   
    geom_line(aes(x=frequency, y=cutoff)))
  print("\n")
  # save dataframe
  invoutliersnps <- invdf[invdf$outlier == TRUE,]
  fwrite(invoutliersnps, paste0('out/inversion/',inv,'_outlier_snps_annotated.txt'), sep='\t')

  invdf <- invdf %>% dplyr::group_by(bin) %>% dplyr::mutate(zrob = robustscale(area))
  
  invdf <- invdf %>% dplyr::group_by(bin) %>% dplyr::mutate(binmean = mean(area), binsd = sd(area))
  invdf <- invdf %>% dplyr::mutate(snpZscore = (area - binmean)/binsd)
  assign(paste0("invdf",inv), invdf)
  
  # hypergeometric tests
  for (ann in unique(invdf$annotation)){
    q <- nrow(invdf[invdf$outlier == TRUE & invdf$annotation == ann,])
    m <- nrow(invdf[invdf$annotation == ann,])
    n <- nrow(invdf[! invdf$annotation == ann,])
    k <- nrow(invdf[invdf$outlier == TRUE,])
    
    print(paste(ann, ': p = ', phyper(q, m, n, k, lower.tail=FALSE)))
  NSoutliergenes <- invdf[invdf$outlier == TRUE & invdf$annotation == 'missense_variant',]$`Gene ID`
  NSoutliergenes <- as.list(unique(NSoutliergenes[NSoutliergenes != 'none']))
  fwrite(NSoutliergenes, paste0('out/inversion/',inv,'_NS_outlier_genes.txt'), sep='\n')
  
  SYoutliergenes <- invdf[invdf$outlier == TRUE & invdf$annotation == 'synonymous_variant',]$`Gene ID`
  SYoutliergenes <- as.list(unique(SYoutliergenes[SYoutliergenes != 'none']))
  fwrite(SYoutliergenes, paste0('out/inversion/',inv,'SY_outlier_genes.txt'), sep='\n')
  
  outliergenes <- invdf[invdf$outlier == TRUE,]$`Gene ID`
  outliergenes <- as.list(unique(outliergenes[outliergenes != 'none']))
  fwrite(outliergenes, paste0('out/inversion/',inv,'outlierSNP_genes.txt'), sep='\n')
  
  hifioutliergenes <- invdf[invdf$outlier == TRUE & invdf$frequency >= 0.75,]$`Gene ID`
  hifioutliergenes <- as.list(unique(hifioutliergenes[hifioutliergenes != 'none']))
  fwrite(hifioutliergenes, paste0('out/inversion/',inv,'hifioutlierSNP_genes.txt'), sep='\n')
  }
}
```

```{r merge-dfs, echo=FALSE}
allsnps <- rbind(allsnps, invdf2La, invdf2Rb, invdf2Rc)
```
  
# Windowed outlier density

Since using just our outlier cutoff gives us a *ton* of SNPs to sort through,
we'll look at the distribution of outliers across the genome using a windowed analysis.
Our test statistic will be a z score that represents the proportion of outliers relative to 
the number of SNPs found within a window. 
To account for regions with low recombination, which will only have a few SNPs and thus could skew our statistic,
we'll mask regions with a recombination rate below 1.5 cM/Mb.

```{r calculate-windows, echo=FALSE}
snphist <- function(snpframe){
  ohs <- hist(snpframe[snpframe$outlier==TRUE,]$position,
              breaks=seq(from=round(min(snpframe$position)), to=max(snpframe$position)+100000, by=100000))
  nhs <- hist(snpframe[snpframe$outlier==FALSE,]$position,
              breaks=seq(from=round(min(snpframe$position)), to=max(snpframe$position)+100000, by=100000))
  breaks <- ohs$breaks[1:length(ohs$breaks)-1]
  df <- data.frame(breaks=breaks, mids=ohs$mids, outliers=ohs$counts, nonoutliers=nhs$counts)
  
  #hs <- hist(snpframe$position, 
            # breaks=seq(from=0, to=max(snpframe$position)+100000, by=100000))
  #breaks <- hs$breaks[1:length(hs$breaks)-1]
  #df <- data.frame(breaks=breaks, mids=hs$mids, count=hs$counts)
  return(df)
}


for (chr in c('2L','2R','3L','3R','X')){
  if (chr == '2L'){
    l2La <- allsnps[allsnps$chromosome==chr & allsnps$INV=='none' & allsnps$position<Lastart,]
    l2LaH <- snphist(l2La)
    l2LaH$INV <- 'none'
    l2LaH$chromosome <- chr
    i2La <- allsnps[allsnps$INV == '2La' & allsnps$position >=Lastart & allsnps$position <= Lastop,]
    i2LaH <- snphist(i2La)
    i2LaH$INV <- '2La'
    i2LaH$chromosome <- chr
    r2La <- allsnps[allsnps$chromosome==chr & allsnps$INV=='none' & allsnps$position>Lastop,]
    r2LaH <- snphist(r2La)
    r2LaH$INV <- 'none'
    r2LaH$chromosome <- chr
    
    assign(paste0('all',chr,'H'), rbind(l2LaH, i2LaH, r2LaH))
  }
  else if (chr == '2R'){
    l2Rb <- allsnps[allsnps$chromosome==chr & allsnps$INV=='none' & allsnps$position<Rbstart,]
    l2RbH <- snphist(l2Rb)
    l2RbH$INV <- 'none'
    l2RbH$chromosome <- chr
    i2Rb <- allsnps[allsnps$INV == '2Rb' & allsnps$position >= Rbstart & allsnps$position <= Rbstop,]
    i2RbH <- snphist(i2Rb)
    i2RbH$INV <- '2Rb'
    i2RbH$chromosome <- chr
    i2Rc <- allsnps[allsnps$INV == '2Rc' & allsnps$position >= Rcstart & allsnps$position <= Rcstop,]
    i2RcH <- snphist(i2Rc)
    i2RcH$INV <- '2Rc'
    i2RcH$chromosome <- chr
    r2Rc <- allsnps[allsnps$chromosome==chr & allsnps$INV=='none' & allsnps$position>Rcstop,]
    r2RcH <- snphist(r2Rc)
    r2RcH$chromosome <- chr
    r2RcH$INV <- 'none'
    
     assign(paste0('all',chr,'H'), rbind(l2RbH, i2RbH, i2RcH, r2RcH))
  }
  else{
    tmpdf <- allsnps[allsnps$chromosome==chr,]
    tmpdf <- snphist(tmpdf)
    tmpdf$INV <- 'none'
    tmpdf$chromosome <- chr
    assign(paste0('all',chr,'H'), tmpdf)
  }
}


```


```{r merge-dfs-1, echo=FALSE}
allsnphist <- rbind(all2LH, all2RH, all3LH, all3RH, allXH)

# calculate outlier statistic

allsnphist$snps <- allsnphist$outliers + allsnphist$nonoutliers
allsnphist$statistic <- NA
for (I in c('none','2La','2Rb','2Rc')){
  Imask <- allsnphist$INV == I
  emp_q <- sum(allsnphist[Imask,]$outliers) / sum(allsnphist[Imask,]$snps)
  st <- ((allsnphist[Imask,]$outliers / allsnphist[Imask,]$snps) - emp_q) / (sqrt((emp_q * (1-emp_q))/allsnphist[Imask,]$snps))
  allsnphist[Imask,]$statistic <- st
}

allsnphist$ends <- allsnphist$breaks + (100000 - 1)
```

```{r rec-map, echo=FALSE}
read_recomb <- function(chromosome){
  df <- fread(paste0('data/relernn_data/agam.',chromosome,'.CAR.rmap.txt.gz'))
  df <- df %>% arrange(end)
  df$start <- 0
  for (i in seq(2, nrow(df))) {
    df[i, 'start'] <- df[i-1,'end']
  }
  df$bp <- df$end - df$start
  return(df)
}

for (c in c('2L','2R','3L','3R')){
  df <- read_recomb(c)
  assign(paste0('r',c), df)
}

allsnphist$mean_cMMb <- 0
for (i in seq(nrow(allsnphist))) {
    st <- allsnphist[i,'breaks'][[1]]
    en <- allsnphist[i,'ends'][[1]]
    c <- allsnphist[i,'chromosome'][[1]]
    if (c=='X') next
    rcdf <- get(paste0('r',c))
    left <- rcdf[(rcdf$end > st & rcdf$end < en & rcdf$start < st),]
    left$bp <- left$end - st
    right <- rcdf[(rcdf$start < en & rcdf$end > en & rcdf$end > st),]
    right$bp <- en - right$start
    overlap <- rbind(left, rcdf[(rcdf$start > st & rcdf$end < en),],right)
    allsnphist[i,]$mean_cMMb <- weighted.mean(overlap$cMMb, overlap$bp)
}

recsnphist <- allsnphist
recsnphist$statistic[recsnphist$mean_cMMb < 1.5] <- NaN
#recsnphist <- allsnphist[allsnphist$mean_cMMb > 1.5,]
recsnphist <- recsnphist[recsnphist$chromosome != 'X',]

#ggplot(recsnphist) + theme_bw() +
#  geom_line(aes(x=mids, y=statistic, color=INV), size=0.5) +  facet_wrap(vars(chromosome))
```

Cool! Now let's set another 90th percentile cutoff:

```{r window-cutoff, echo=FALSE}
recsnphist <- recsnphist %>% group_by(INV) %>% mutate(Q = quantile(statistic, 0.9, na.rm=TRUE)) #%>% ungroup()

ggplot(recsnphist) + theme_bw() +
  geom_line(mapping=aes(x=mids, y = statistic, color=INV), size=0.5) +  
  geom_hline(mapping=aes(yintercept=Q, color=INV)) +
  facet_wrap(vars(chromosome))
```

```{r tag-outliers, echo=FALSE}
recsnphist <- recsnphist[!is.na(recsnphist$statistic),]
recsnphist$nstat_outlier <- FALSE
recsnphist[recsnphist$statistic > recsnphist$Q,]$nstat_outlier <- TRUE

allsnps$nstat_outlier <- FALSE
allsnps$nstat <- 0
for (i in seq(nrow(recsnphist))){
  allsnps[(allsnps$chromosome == recsnphist[i,]$chromosome &
         allsnps$position >= recsnphist[i,]$breaks &
         allsnps$position <= recsnphist[i,]$ends),]$nstat <- recsnphist[i,]$statistic
  if (recsnphist[i, 'nstat_outlier'] == TRUE) { 
    allsnps[(allsnps$chromosome == recsnphist[i,]$chromosome &
             allsnps$position >= recsnphist[i,]$breaks &
             allsnps$position <= recsnphist[i,]$ends &
             allsnps$outlier == TRUE),]$nstat_outlier <- TRUE
  }
}
```

```{r enrichment-test-windows, echo=FALSE}
for (inversion in c('none','2La','2Rc','2Rb')){
  print(inversion)
  for (ann in unique(allsnps$annotation)){
    q <- nrow(allsnps[allsnps$nstat_outlier == TRUE & allsnps$annotation == ann & allsnps$INV == inversion,])
    m <- nrow(allsnps[allsnps$annotation == ann & allsnps$INV == inversion,])
    n <- nrow(allsnps[! allsnps$annotation == ann & allsnps$INV == inversion,])
    k <- nrow(allsnps[allsnps$nstat_outlier == TRUE & allsnps$INV == inversion,])
    
    print(paste(ann, ': p = ', phyper(q, m, n, k, lower.tail=FALSE)))
  }
}
```

```{r significance}
# adding p values for our Z scores!
allsnps$snpPval <- pnorm(q=allsnps$snpZscore, lower.tail=TRUE)
allsnps$windowPval <- pnorm(q=allsnps$nstat, lower.tail=FALSE)
```


```{r save-final}
fwrite(allsnps, 'out/all_snps_genome_scan_annotated.txt')
fwrite(allsnphist, 'out/all_snps_windowed_genome_scan_annotated.txt')
fwrite(recsnphist, 'out/high_recombination_windowed_genome_scan_annotated.txt')
```

```{r read-final}
allsnps <- fread('out/all_snps_genome_scan_annotated.txt')
allsnphist <- fread('out/all_snps_windowed_genome_scan_annotated.txt')
recsnphist <- fread('out/high_recombination_windowed_genome_scan_annotated.txt')
```

## How does this statistic compare to other scans for selection?

### IHS

IHS is commonly used to ID instances of selection...

```{r read-ihsc}
# code block for reading in ihs tsv files
rf <- function(filepath){
  df <- fread(filepath)
  windowsize <- strsplit(filepath, '_')[[1]][7] %>% str_remove('kb') %>% as.numeric()
  df$start <- round(df$window - windowsize/2)
  df$stop <- round(df$window + windowsize/2)
  return(df)
}
ihs <- rbindlist(lapply(list.files('data/ihs12/', full.names = TRUE), rf), fill=TRUE)
```

```{r compare-ihsc}
allsnphist$IHS2 <- 0
for (i in seq(nrow(allsnphist))) {
    st <- allsnphist[i,'breaks'][[1]]
    en <- allsnphist[i,'ends'][[1]]
    c <- allsnphist[i,'chromosome'][[1]]
    left <- ihs[(ihs$stop > st & ihs$stop < en & ihs$start < st),]
    left$bp <- left$stop - st
    middle <- ihs[(ihs$start > st & ihs$stop < en),]
    middle$bp <- middle$stop - middle$start
    right <- ihs[(ihs$start < en & ihs$stop > en & ihs$stop > st),]
    right$bp <- en - right$start
    overlap <- rbind(left, middle ,right)
    allsnphist[i,]$IHS2 <- weighted.mean(overlap$h12, overlap$bp)
}

ggplot(allsnphist) + geom_point(aes(x=statistic, y=IHS2)) +
  geom_smooth(aes(x=statistic, y=IHS2), method='lm')

print(summary(lm(IHS2~statistic, allsnphist)))
```

### diploSHIC

```{r read-shic}
# code block for reading in shic bed files
rf <- function(filepath){
  df <- fread(filepath, header=FALSE)
  colnames(df) <- c('chromosome','start','stop','class','x','y','start','stop','params')
  df$class <- str_split_fixed(df$class, pattern = '_', n=3)[,1]
  return(df)
}

mosqs <- c('data/partialSHIC/empirical_mosquito_sweep_classifications/BFM.',
           'data/partialSHIC/empirical_mosquito_sweep_classifications/BFS.',
           'data/partialSHIC/empirical_mosquito_sweep_classifications/CMS.',
           'data/partialSHIC/empirical_mosquito_sweep_classifications/GNS.')
chrms <- c('2L.bed','2R.bed','3L.bed','3R.bed')
shic <- rbindlist(lapply(expand.grid(mosqs, chrms) %>% apply(1 , paste , collapse = ""), rf), fill=TRUE)

# mark shic outputs
sweep_classes <- c('SoftPartial',
                   'HardPartial',
                   'Soft',
                   'Hard')
shic$sweep <- 0
shic[(shic$class %in% sweep_classes),'sweep'] <- 1

shic$chromosome <- shic$chromosome %>% str_remove('chr')
```

```{r compare-shic}
allsnphist$SHIC <- 0
for (i in seq(nrow(allsnphist))){
  wd <- allsnphist[i,]
  sp <- sum(shic[shic$chromosome == wd$chromosome & shic$start >= (wd$breaks) & shic$stop <= (wd$breaks + 1e5),]$sweep)
  allsnphist[i,'SHIC'] <- sp
}

ggplot(allsnphist) + geom_point(aes(x=statistic, y=SHIC)) +
  geom_smooth(aes(x=statistic, y=SHIC), method='lm')

print(summary(lm(SHIC~statistic, allsnphist)))
```

# Functional consequences

Are the genes that outlier SNPs are in enriched for any sorts of functions? I'll output Ensembl gene IDs so I can run DAVID for enrichment tests:

```{r outlier-snp-genes}
NSoutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$annotation == 'missense_variant',]$`Gene ID`
NSoutliergenes <- as.list(unique(NSoutliergenes[NSoutliergenes != 'none']))
fwrite(NSoutliergenes, 'out/all_NS_outlier_genes.txt', sep='\n')

SYoutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$annotation == 'synonymous_variant',]$`Gene ID`
SYoutliergenes <- as.list(unique(SYoutliergenes[SYoutliergenes != 'none']))
fwrite(SYoutliergenes, 'out/all_SY_outlier_genes.txt', sep='\n')

outliergenes <- allsnps[allsnps$outlier == TRUE,]$`Gene ID`
outliergenes <- as.list(unique(outliergenes[outliergenes != 'none']))
fwrite(outliergenes, 'out/all_outlierSNP_genes.txt', sep='\n')

hifioutliergenes <- allsnps[allsnps$outlier == TRUE & allsnps$frequency >= 0.75,]$`Gene ID`
hifioutliergenes <- as.list(unique(hifioutliergenes[hifioutliergenes != 'none']))
fwrite(hifioutliergenes, 'out/all_hifioutlierSNP_genes.txt', sep='\n')
```

```{r windowoutlier-snp-genes}
wNSoutliergenes <- allsnps[allsnps$nstat_outlier == TRUE & allsnps$annotation == 'missense_variant',]$`Gene ID`
wNSoutliergenes <- as.list(unique(wNSoutliergenes[wNSoutliergenes != 'none']))
fwrite(wNSoutliergenes, 'out/all_wNS_outlier_genes.txt', sep='\n')

wSYoutliergenes <- allsnps[allsnps$nstat_outlier == TRUE & allsnps$annotation == 'synonymous_variant',]$`Gene ID`
wSYoutliergenes <- as.list(unique(wSYoutliergenes[wSYoutliergenes != 'none']))
fwrite(wSYoutliergenes, 'out/all_wSY_outlier_genes.txt', sep='\n')

woutliergenes <- allsnps[allsnps$nstat_outlier == TRUE,]$`Gene ID`
woutliergenes <- as.list(unique(woutliergenes[woutliergenes != 'none']))
fwrite(woutliergenes, 'out/all_woutlierSNP_genes.txt', sep='\n')

whifioutliergenes <- allsnps[allsnps$nstat_outlier == TRUE & allsnps$frequency >= 0.75,]$`Gene ID`
whifioutliergenes <- as.list(unique(whifioutliergenes[whifioutliergenes != 'none']))
fwrite(whifioutliergenes, 'out/all_whifioutlierSNP_genes.txt', sep='\n')
```


To broaden our search to non-genic SNPs, I'll look at annotations for outlier regions of the genome...

```{r gff}
library(stringi)
ann <- fread('data/VectorBase-68_AgambiaePEST.gff')
colnames(ann) <- c('seqname', 'source','feature','start','end','score','strand','frame','attribute')
#ann$ID <- 
atts <- strsplit(ann$attribute, split = ';')
IDs <- lapply(atts, function(l) l[[1]] %>% str_replace("ID=",''))
ann$ID <- unlist(IDs)

descs <- lapply(atts, function(l) l[stri_detect_fixed(l, "description=")] %>% str_replace("description=",''))
ann$description <- descs
ann$ends <- ann$end
ann$starts <- ann$start
recsnphist[recsnphist$nstat_outlier,c('chromosome','breaks','ends')]

```

```{r interactive-plotly-nstatoutlier}
library(plotly)

p <- ggplot() + theme_bw() + 
    geom_point(data=allsnps[allsnps$nstat_outlier == TRUE,], aes(text=paste0('gene: ',`Gene ID`,'\n','description: ',`Product Description`), x=frequency, y=area, color=annotation, size=nstat, group=`Gene ID`)) + scale_size_continuous(range=c(0.1, 2), trans='log10') + #scale_alpha_manual(values=c(0.25, 0.5)) +
  theme(legend.position='none')

ggplotly(p, width=1000, height=700)
```

```{r interactive-plotly-outlier}
library(plotly)

p <- ggplot() + theme_bw() + 
    geom_point(data=allsnps[allsnps$outlier == TRUE & allsnps$nstat_outlier == FALSE,], aes(text=paste0('gene: ',`Gene ID`,'\n','description: ',`Product Description`), x=frequency, y=area, color=annotation, size=nstat, group=`Gene ID`)) + scale_size_continuous(range=c(0.1, 2), trans='log10') + #scale_alpha_manual(values=c(0.25, 0.5)) +
  theme(legend.position='none')

ggplotly(p, width=1000, height=700)
```

# IR SNPs

## VGSC

```{r vgsc}
library(DT)


# The locus occurs on chromosome arm 2L from position 2,358,158 to 2,431,617.
chr <- '2L' # chromosome
ist <- 2358158 # locus start
isp <- 2431617 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```

## GSTE

```{r gste}
#The locus occurs on chromosome arm 3R from position 28,591,663 to 28,602,280.
chr <- '3R' # chromosome
ist <- 28591663 # locus start
isp <- 28602280 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```

## Cyp6p

```{r cyp6p}
# The locus occurs on chromosome arm 2R from position 28,480,576 to 28,505,816.

# pushing back start to include Solute Carrier Family 8
chr <- '2R' # chromosome
ist <- 28420677 # locus start
isp <- 28505816 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```

## Ace1

```{r ace1}
# The locus occurs on chromosome arm 2R from position 3,484,107 to 3,495,790.
chr <- '2R' # chromosome
ist <- 3484107 # locus start
isp <- 3495790 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```

## Rdl

```{r rdl}
# The locus occurs on chromosome arm 2L from position 25,363,652 to 25,434,556.
chr <- '2L' # chromosome
ist <- 25363652 # locus start
isp <- 25434556 # locus stop


DT::datatable((allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),]))
```

## Cyp9k1

```{r cyp9k1}
# The locus occurs on chromosome X from position 15,240,572 to 15,242,864.
chr <- 'X' # chromosome
ist <- 15240572 # locus start
isp <- 15242864 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```

## Tep1

```{r tep1}
# The locus occurs on chromosome arm 3L from position 11,202,091 to 11,206,882.
chr <- '3L' # chromosome
ist <- 11202091 # locus start
isp <- 11206882 # locus stop


DT::datatable(allsnps[allsnps$chromosome==chr &
          allsnps$position >= ist &
          allsnps$position <= isp &
          (allsnps$nstat_outlier | allsnps$outlier),])
```



```{r}
library(DT)

DT::datatable(allsnps[allsnps$outlier==TRUE,], class='cell-border stripe', rownames=F, filter='top',
              editable=TRUE, extensions='Buttons', options=list(
                dom='Bfrtip', buttons=c('copy','csv','pdf')
              ))
```


# Nonsynonymous SF outliers

Lastly, let's use ThermoMPNN to investigate if any nonsynonymous SF outliers have impacts on protein stability. 
I used VectorBase to get UniProt IDs for each gene containing nonsyn SF outliers, then downloaded their PDB files from AlphaFold and ran ThermoMPNN to do deep mutational scanning on each protein.
From the MalariaGen Python API, I also grabbed the specific amino acid change associated with each of these nonsynonymous SNPs (using `ag1000g-altalleles.ipynb`), which I'll use to line up ThermoMPNN results with our outliers.

Here's all the nonsynonymous SF outlier SNPs that change protein stability:

```{r write-uniprots, echo=FALSE}
df <- fread('out/nonsynonymous-outliers-uniprot.txt')
uniprots <- unique(strsplit(paste0(df$`UniProt ID(s)`, collapse=','), ',')[[1]])
uniprots <- uniprots[uniprots != 'N/A']
uniprots <- as.list(uniprots)
fwrite(uniprots, 'out/missense_outlier_snp_uniprotID.txt', sep='\n')
```

```{r merge-thermo-data, echo=FALSE}
up <- df %>% separate(col='UniProt ID(s)', into='UniProt ID(s)', sep=',')
missense_df <- merge(up, allsnps[allsnps$outlier & allsnps$annotation == 'missense_variant',])
missense_df <- missense_df[missense_df$`UniProt ID(s)` != 'N/A',]
snp_alleles <- fread('out/outlier_snps_ancestral_derived.txt')
snp_alleles <- snp_alleles %>% dplyr::select(-c(V1))
snp_alleles$source_id <- snp_alleles$source_id_x
missense_df <- merge(missense_df, snp_alleles[,c('source_id', 'ancestral_allele', 'derived_allele', 'AA_change','position','alternate','chromosome')])
rf <- function(filepath){
  df <- fread(filepath)
  df$`UniProt ID(s)` <- strsplit((strsplit(filepath, '/')[[1]][3] %>% str_remove('.csv')), '-')[[1]][2]
  df$AA_change <- df$Mutation
  return(df)
}

thermo <- rbindlist(lapply(list.files('out/thermompnn-out', full.names=TRUE), rf), fill=TRUE)

missense_df <- left_join(missense_df, thermo, by=c('UniProt ID(s)' = 'UniProt ID(s)', 'AA_change' = 'AA_change'))
missense_df <- missense_df[!is.na(missense_df$`ddG (kcal/mol)`) & missense_df$`ddG (kcal/mol)` != 0,]
DT::datatable(missense_df[,c('Gene ID', 'Product Description', 'frequency', 'area', 'ddG (kcal/mol)', 'AA_change', 'nstat_outlier', 'nstat', 'snpZscore', 'zrob', 'snpPval')] %>% arrange(-`ddG (kcal/mol)`), class='cell-border stripe', rownames=F, filter='top',
              editable=TRUE, extensions='Buttons', options=list(
                dom='Bfrtip', buttons=c('copy','csv','pdf')
              ))

fwrite(missense_df, 'out/missense_snps_stabilitychange.txt', sep='\t')
```


I'm gonna try to manually annotate these guys now by merging them with their GO terms (from VectorBase) and hand-downloaded IR candidate genes from the MalariaGen selection atlas...

```{r}
df <- fread('out/nonsynonymous_vectorbase_goterms.txt')
missense_df <- missense_df[,c('source_id',
                              'position',
                              "alternate",
                              "chromosome",
                              "Gene ID",
                              "Product Description",
                              "Genomic Location (Gene)",
                              "Gene Type",              
                              "UniProt ID(s)",
                              "Gene Name or Symbol",
                              "gene_source_id",
                              "frequency",
                              "annotation",
                              "impact",
                              "area",
                              "locs",                
                              "maxdist",
                              "INV",
                              "bin",
                              "bin_frequency",
                              "Q",
                              "cutoff",
                              "outlier",
                              "nstat_outlier",
                              "nstat",
                              "ancestral_allele",
                              "derived_allele",
                              "AA_change",
                              "Mutation",
                              "ddG (kcal/mol)",
                              "pos",
                              "wtAA",
                              "mutAA")]

missense_go <- full_join(df, missense_df)
irdf <- rbind(fread('data/cuticular-IRcandidates.tsv'),
              fread('data/behavioral-IRcandidates.tsv'),
              fread('data/ts-IRcandidates.tsv'),
              fread('data/metabolic-IRcanidates.tsv'))
irdf$`Gene ID` <- irdf$Gene
missense_ir <- left_join(missense_go, irdf)
#fwrite(missense_ir, 'out/missense_goterms_manual.txt', sep='\t')
missense_ir <- fread('data/missense_goterms_manual.tsv', sep='\t') %>% dplyr::select(-(last_col()))

missense_ir
```

```{r}
DT::datatable(missense_ir %>% arrange(-`ddG (kcal/mol)`), class='cell-border stripe', rownames=F, filter='top',
              editable=TRUE, extensions='Buttons', options=list(
                dom='Bfrtip', buttons=c('copy','csv','pdf')
              ))
```


# Vignettes

## AGAP029542

Do any of these variants change splice sites?

```{r make-snpAI-annotation}
gff <- fread('data/VectorBase-68_AgambiaePEST.gff')
colnames(gff) <- c('CHROM','SOURCE','TYPE','START','STOP','X','STRAND','Y','INFO')

gff <- gff[gff$CHROM == 'AgamP4_3R',]
proteins <- gff[gff$TYPE == 'protein_coding_gene',]
mrnas <- gff[gff$TYPE == 'mRNA',]
mrnas <- mrnas %>% tidyr::separate(INFO, sep=';', into=c('ID','Parent','INFO'), extra='merge')
mrnas$ID <- str_remove(mrnas$ID, 'ID=')
mrnas$Parent <- str_remove(mrnas$Parent, 'Parent=')
exons <- gff[gff$TYPE == 'exon',]
exons <- exons %>% tidyr::separate(INFO, sep=';', into=c('ID', 'Parent', 'INFO'), extra='merge')
exons$ID <- str_remove(exons$ID, 'ID=')
exons$Parent <- str_remove(exons$Parent, 'Parent=')


ann3R <- merge(mrnas, exons,
      by.x='ID', by.y='Parent',
      suffixes=c('.mrna','.exon'))
exons3R <- ann3R %>% dplyr::group_by(ID) %>% dplyr::summarize(
                                                    EXON_START = na_if(coalesce(str_c(START.exon, collapse = ","),
                                                                str_c(START.exon[complete.cases(START.exon)], collapse = ",")), ""),
                                                    EXON_END = na_if(coalesce(str_c(STOP.exon, collapse = ","),
                                                                str_c(STOP.exon[complete.cases(STOP.exon)], collapse = ",")), ""))
exons3R$EXON_START <- paste0(exons3R$EXON_START, ',')
exons3R$EXON_END <- paste0(exons3R$EXON_END, ',')

anndf <- merge(mrnas, exons3R)
anndf$NAME <- anndf$ID
anndf$TX_START <- anndf$START
anndf$TX_END <- anndf$STOP

fwrite(anndf[,c('NAME','CHROM','STRAND','TX_START','TX_END','EXON_START','EXON_END')], 'data/3RspliceAI.txt', sep='\t')
```

## Gustatory receptors

Which transcripts are changed?

```{r}
snp_alleles <- fread('out/outlier_snps_ancestral_derived.txt')
grps <- snp_alleles[grepl('gustatory', snp_alleles$`Product Description`, fixed=TRUE) & 
                     snp_alleles$AA_change != '',]
grps$source_id <- grps$source_id_x
grps <- grps %>% dplyr::select(-c(V1, source_id_x, source_id_y))

grps <- merge(grps, missense_df, by='source_id')
DT::datatable(grps, class='cell-border stripe', rownames=F, filter='top',
              editable=TRUE, extensions='Buttons', options=list(
                dom='Bfrtip', buttons=c('copy','csv','pdf')
              ))
```

Get the changed transcripts for AlphaFold...

```{r}
# define transcripts (sourced from Vectorbase)
AGAP006143_RC <- 'MHELQFLNIFNYCFVSPTYLHFDSLCNRYEFQHKNVFRNVTLLIVLVTCSVVSAIAVIVEFMQTSLLSVIGASSIILYTTRFLVMIPLTLWVWLCRHQLISDNNVALNIIQHKHAKFYPMPARRRTLQIYWIQVGSIVILLTAALAGQCITWWEIVVTRKEYFLLNVCALVALEILTLMHRMYIQCWALVIAHHLDELVGFVKNKRVELHLLRAAIVFWEELEFFKQRATSTFRVMNVLHVLDILITCVVETYTIFYVFEMGLSLVDAMLNVITLTVYAATFFMFAYAHDLVKVKEAELKDALKSMQYTNLKRQSRDQKDFYDLVNLKLMMESPKITACGLFEINLQIFYNVFAAIITYIVILFQFRGFEKSP'
AGAP006877_RA <- 'MDPIKLDTTFYHKLLVKKFNLLLNIAQLAGFLPFPAYVLRQDVERLHFRTLVLTCLNLTFAVLIFLTSFVCYLAMYIYYPDLMYKENLPAVLQIMYHVENWLRVVMVLIALVGPRLSGRYFRETIDTLVHIMKLFDRATKIESILTAISVITNRLLLLYGLHALIITVTVWISTEHPVSTLLNVSYLAPYVTIAVYILLYRALLASIAGIVGCLNDNLREITIQDRIDPRRTYGKHTTISYIMLQEGKERVTSPLDVATIVRLSTMHMALMRLARAANKHFGVLMLIIVLSTFIQINMLLLELYHNISHPVMPEYCLWVLFLHAIVHFTFFFVIATSNHAIQQENERTMLLLHEFKCSWSSEQNMTIEHFISQISNLHDVHQACGMLNLDMKLISNAVAAITSIMVVLIQFSDTGL'
AGAP006877_RB <- 'MAWNDTEIDRLRHQLRQNISPTVRLSQCLSLAPYPLSVFQRNCSTRSVRIRIIICFRYAFATCVTVAVVASQFTMFYYFPHIMYQPKVPIFIVILYYIVSILQTLTTGNMMIGCEQRRAEYEGYFEEVLHLMKETAHQPDCKTTIWYRHVTKVLLALYCIASLTVPIVMTTILWDIATIPYVMAQTVPFVVSSLILNQYFCVFVHLTSILRKMNERLARFLNMLPGSATEPPVQLRGKPLIYNVLGESLKMESSRLDQLEQLRLLHVRTVQTAGSLSEKFGIVIILIVIAAFASVNIELLEFYQSIKLGTLTPTTIFMKFLYAASKFSFYILIAYPNRLIQQENQKALFMLYRIKRISCSVELNEAIEHFISQISNLHDVHQACGMLNLDMKLISNAVAAITSIMVVLIQFSDTGL'
AGAP009805_RI <- 'MSNLLKRYRLLLAVASIGYLIPCSYNIRTGLFDCSYRNTLVCVCNVVFFGGFVWYDFGMILKFYATLPIVLVGILTVDVTVYNLLIFCIIINAVYNRDCFVQLLNSLFARDDWMLESVAMQGSSQQRRSTQSTGGLVCLIVLVLLYAMYNALFVNDHSMVLMDMIILLRFCFMFLILELYRVCVRIIRKRMKQLQVLLTQMEEINTTACVEHVVHVFLDRFQRYYLLIDSVNKCFSVPVTHTLLLIVLERTVAAYDVFENLRGESKMILWDFYRLLYRQVWEITYIVLMVLLAINCNATSLQVEETALCTRHFDDYRLQNTRAAKQIQNFLLKNLHQKKKFSACGFFDIDNTVIYMVFSSIVTYLVILIQFKQLETDLTQAGDGYNVTSNVSTVQP'
AGAP009857_RA <- 'MQFSHQHSSDVHFIVSRERSSPSTCTMNLVLPLLPNGQQLLSVAFGIFKCFGFIPFPFDCCTFALTPCSNVRSLLHLPILQVSFYLTLFCIIMSNRVNFFFTGLQILSLNDIIKYGTLTLSVFAIFIDTVLQRNTHRLVWEKIALIRLSARKVYVERFTRHYLWKFYGYLAVCAFLEAQVLYLAWDDPSALAYWLVIMILHAFLRLRHLFHMFFIDILKIHLQKLHHDLVDAGEYMADLVEQPQDTAVFRAMYAQSVDRLLSLKSVYGQLWELSDCINRNFGWSQICNFTGNFVQLSCDLYWLYMSMKWFEATEYKVVIVITLLPSTSIIVLLLSSAESCLGVAASLQSALLDIPMGNDSTFRKIIYRFGLQIAQQRIRLTAHGLFEINYSLLKMFGTGITTYMIIFITFSKDIKLEDIDDE'
AGAP011915_RA <- 'MSKWINKRKVSILNLAPPSKPGPALQATPVGEQEFEQLFHFAFKCFRLFALTPGLMDRQKDRYRVRNTRWMMLIVLLLVVVAWIALFETFFIERRTALITGIANHIQFLMNTIALTVAWVVPQLKADELGSILDGFLLIDRELSSYNVHEVAGYKRVSFLLRYGVVLLALMSLTVYDGFVSFVQLTTVEVWYWLSHQLPFIIYAMAFLHAYVLIYWLHARFRRLNTLVEQYYRQGHIFAPARQTIISFATMVKLDEESAVGEELHSVGRREISDDLQVLAIVSRTIDLGQKIESYFGPLFLTVYTALFSVTTIQSYYCYLHLTAKGDRGLSIETLVLSGGIILYNVIAIVALPYICEQVESESKLLMSYLSKLSMKHSQVAQHSSIWFPNLISSVRFSAFGFFTINYNMLSGLVAGMVTYLIIFIQFNSMVPAGKDDTHHTRKQHVTEERF'

for (i in seq(nrow(grps))){
  print(as.character(grps[i, 'source_id']))
  pseq <- grps[i, 'source_id'] %>% str_replace('-','_') %>% get()
  ancAA <- as.character(grps[i, 'wtAA'])
  derAA <- as.character(grps[i, 'mutAA'])
  posAA <- as.integer(grps[i, 'pos'])
  
  if (substr(pseq, posAA, posAA) == ancAA){
    substr(pseq, posAA, posAA) <- derAA
    print(pseq)
  }
  else {print('oops')}
}




```


